include/watchdog.h

```diff
@@ -55,62 +70,234 @@ struct wdt_config {
 	enum wdt_mode mode;
 	void (*interrupt_fn)(struct device *dev);
 };
+/**
+ * @}
+ */
+
+
+/**
+ * @brief Pause watchdog timer when CPU is in sleep state.
+ */
+#define WDT_OPT_PAUSE_IN_SLEEP	BIT(0)
+
+/**
+ * @brief Pause watchdog timer when CPU is halted by the debugger.
+ */
+#define WDT_OPT_PAUSE_HALTED_BY_DBG	BIT(1)
+
+/**
+ * @brief Watchdog reset flag bit field mask shift.
+ */
+#define WDT_FLAG_RESET_SHIFT		(0)
+/**
+ * @brief Watchdog reset flag bit field mask.
+ */
+#define WDT_FLAG_RESET_MASK		(0x3 << WDT_FLAG_RESET_SHIFT)
+
+/**
+ * @name Watchdog Reset Behavior.
+ * Reset behavior after timeout.
+ * @{
+ */
+/** No reset */
+#define WDT_FLAG_RESET_NONE		(0 << WDT_FLAG_RESET_SHIFT)
+/** CPU core reset */
+#define WDT_FLAG_RESET_CPU_CORE		(1 << WDT_FLAG_RESET_SHIFT)
+/** Global SoC reset */
+#define WDT_FLAG_RESET_SOC		(2 << WDT_FLAG_RESET_SHIFT)
+/**
+ * @}
+ */
+
+/**
+ * @brief Watchdog timeout window.
+ *
+ * Each installed timeout needs feeding within the specified time window,
+ * otherwise the watchdog will trigger. If the watchdog instance does not
+ * support window timeouts then min value must be equal to 0.
+ *
+ * @param min Lower limit of watchdog feed timeout in milliseconds.
+ * @param max Upper limit of watchdog feed timeout in milliseconds.
+ *
+ * @note If specified values can not be precisely set they are always
+ *	 rounded up.
+ */
+struct wdt_window {
+	u32_t min;
+	u32_t max;
+};
+
+/** Watchdog callback. */
+typedef void (*wdt_callback_t)(struct device *dev, void *handle);
 
-typedef void (*wdt_api_enable)(struct device *dev);
-typedef void (*wdt_api_disable)(struct device *dev);
-typedef int (*wdt_api_set_config)(struct device *dev,
-				  struct wdt_config *config);
-typedef void (*wdt_api_get_config)(struct device *dev,
-				   struct wdt_config *config);
-typedef void (*wdt_api_reload)(struct device *dev);
+/**
+ * @brief Watchdog timeout configuration struct.
+ *
+ * @param window Timing parameters of watchdog timeout.
+ * @param handle Unique handle used to feed specific watchdog timeout.
+ *		 It is allowed to use the same handle for multistage timeouts.
+ *		 This handle is filled by the user, and will be supplied to
+ *		 callback. A good example is a handle pointing to the id of the
+ *		 thread that is responsible for feeding the corresponding
+ *		 timeout.
+ * @param callback Timeout callback. Passing NULL means that no callback
+ *		   will be run.
+ * @param next Pointer to the next timeout configuration. This pointer is used
+ *	       for watchdogs with staged timeouts functionality. Value must be
+ *	       NULL for single stage timeout.
+ * @param flags Bit field with following parts:
+ *
+ *     reset        [ 0 : 1 ]   - perform specified reset after timeout/callback
+ */
+struct wdt_timeout_cfg {
+	struct wdt_window window;
+	void *handle;
+	wdt_callback_t callback;
+#ifdef CONFIG_WDT_HAS_MULTISTAGE
```

Comment:
I believe we need also `WDT_MULTISTAGE` Kconfig option. It is very likely that an application written for a single stage watchdog when used with multi-stage one will fail to initialize the `next` field to NULL. This may result in a tricky to find bug. It will be cleaner if we use a standard HAVE_* pattern as described in [Kconfig language manual](https://www.kernel.org/doc/Documentation/kbuild/kconfig-language.txt).

`WDT_MULTISTAGE` should be a visible Kconfig option which an application has to explicitly enable. It should depend on `WDT_HAS_MULTISTAGE`. All multistage watchdogs should select `WDT_HAS_MULTISTAGE`. Thus the `WDT_MULTISTAGE` will be visible only when a multi-stage watchdog is enabled.

This line should be changed to
```
#ifdef CONFIG_WDT_MULTISTAGE
```

Reply:
@mnkp What if watchdog supports only multistage timeouts? Then `CONFIG_WDT_MULTISTAGE` would have to be forced to `y`, is there any way to block user from changing it to `n`?

Reply:
Multistage watchdogs will have to support single stage mode. This is always possible. This way application which was written for a single-stage watchdog can be used with a multi-stage one. 

We should think of `WDT_MULTISTAGE` as an optional feature which can be supported only by multi-stage watchdogs.